keps/sig-testing/1746-stage-e2e-test-framework/README.md

```diff
@@ -0,0 +1,610 @@
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+I, as a core Kubernetes contributor, want a e2e test framework that is well
+organized and containsa core of functions that will help me write clean and
+maintainable e2e tests.
+
+#### Story 2
+
+I, as a maintainer of a project running on top of Kubernetes (e.g. external cloud providers, CSI drivers), want to use the
+same infrastructure and tools used by core Kubernetes without having to import
+core Kubernetes.
+
+### Managing Internal Kubernetes Packages
+
+To decouple the e2e test framework from internal packages we essentially have
+to prevent any from `k8s.io/kubernetes/pkg` and from any other package in
+`k8s.io/kubernetes/test` that aren't under
+`k8s.io/kubernetes/test/e2e/framework`.
+
+These core dependencies are introduced in the following ways
+1. as primitives for test utility functions to manage Kubernetes resources and objects
+2. as data structures to manage and consume data generated by a Kubernetes component
+3. as variables and constants
+
+To resolve the dependencies introduced by bullet point 1, we plan to move or
+rewrite e2e test framework functions using official packages and APIs to match
+the current desired behaviour.
+For the most part this should be straightforward and it will only cause issues
+when one of the other two following conditions are present.
+
+Bullet point number 2 is present in cases where e2e tests involve collecting
+metrics from unpublished APIs.
+As an example, see
+[kubeletstatsv1alpha1](https://github.com/kubernetes/kubernetes/blob/ba3bf32300574d86c98657981c96ca609de787a2/test/e2e/framework/resource_usage_gatherer.go#L37)
+being imported to collect Kubelet statistics.
+We want to prevent this kind of internal dependencies in order to ensure that
+all artifacts needed for e2e testing are properly surfaced to users.
+To deal with these cases we will collaborate with the owning SIGs to have the
+necessary internal APIs used in e2e tests published in
+https://github.com/kubernetes/api.
+
+Bullet point 3 will be dealt in a similar manner by working with owning SIGs to
+export any constants and variables defined in core Kubernetes that are required for e2e tests.
+
+### Managing Test Packages
+
+Decoupling the e2e test framework from other test packages will proceed in a
+similar manner.
+
+We will take a look at the utility functions surfaced by the e2e test framework
+and we will see how and why they are used in e2e tests.
+If we encounter utilities that are used sparingly in only a single test package
+then we will opt to move it to that test package.
+
+On the other hand, if we see a lot of dependencies and complexity introduced by
+requiring other test packages then we will work to migrate the essential parts
+within the e2e test framework.
+
+Ultimately, we want a library that is well encapsulated and has useful methods
+for maintainers to write e2e tests.
```

Comment:
Is there a reason this is not better served starting from a library that is built with this in mind rather than introducing churn into our existing code?

Reply:
There are libaries like ginkgo which are for running e2e tests, but the libraries that are part of the e2e test framework are utility functions for doing tasks in e2e tests.
For example, functions that run a specific kind of pod and test conditions on it.
These are the libraries we want to maintain.
We want to maintain and clean up the code that people use as opposed to proposing a library to run e2e tests in some other way.